### DISCUSSION TITLE
Nodes are not visualized in Theia

### DISCUSSION THREAD

**Discussion by hsd-dev** [eclipse-glsp/glsp#667] [Q&A] [answered]
Sorry for spamming the discussion thread. Now I am trying to setup the application with my own Ecore meta-model. Looks like the nodes are created successfully and GModel is also getting updated accordingly. 
Below is the model which started from empty but now has 4 links:
```
<?xml version="1.0" encoding="UTF-8"?>
<model:Links xmi:version="2.0" xmlns:xmi="http://www.omg.org/XMI" xmlns:model="http://www.ipa.fraunhofer.de/model" name="test">
  <body>
    <link resolved="NewLink0"/>
    <link resolved="NewLink1"/>
    <link resolved="NewLink2"/>
    <link resolved="NewLink3"/>
  </body>
</model:Links>
```
and the notation model:
```
<?xml version="1.0" encoding="ASCII"?>
<notation:Diagram xmi:version="2.0" xmlns:xmi="http://www.omg.org/XMI" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:notation="http://www.eclipse.org/glsp/notation">
  <semanticElement elementId="root"/>
  <elements xsi:type="notation:Shape">
    <semanticElement elementId="NewLink0"/>
    <position x="338.0" y="226.0"/>
    <size width="60.0" height="25.0"/>
  </elements>
  <elements xsi:type="notation:Shape">
    <semanticElement elementId="NewLink1"/>
    <position x="591.0" y="248.0"/>
    <size width="60.0" height="25.0"/>
  </elements>
  <elements xsi:type="notation:Shape">
    <semanticElement elementId="NewLink2"/>
    <position x="550.0" y="421.0"/>
    <size width="60.0" height="25.0"/>
  </elements>
  <elements xsi:type="notation:Shape">
    <semanticElement elementId="NewLink3"/>
    <position x="266.0" y="149.0"/>
    <size width="60.0" height="25.0"/>
  </elements>
</notation:Diagram>
```
But I don't see the nodes in the editor. What else is needed to visualize it? Note that my model does not have an id element. 

The code for `CreateLinkNodeHandler`
```
  Link newLink = createLink();
  Command linkCommand = AddCommand.create(editingDomain, model.getBody(), ModelPackage.Literals.BODY__LINK, newLink);

  Shape shape = createShape(newLink.getResolved(), relativeLocation);
  Command shapeCommand = AddCommand.create(editingDomain, diagram, NotationPackage.Literals.DIAGRAM__ELEMENTS, shape);

  CompoundCommand compoundCommand = new CompoundCommand();
  compoundCommand.append(taskCommand);
  compoundCommand.append(shapeCommand);
  return compoundCommand;
```

**Comment by hsd-dev** [comment]
In `fillRootElement` in print out 
```
System.out.println(notationModel.getSemanticElement());
System.out.println(notationModel.getSemanticElement().getResolvedSemanticElement());
```
and the output is
```
org.eclipse.glsp.server.emf.model.notation.impl.SemanticElementReferenceImpl@62cac08f (elementId: root)
null
```
Both the .tasklist and .notation files pasted above are hand-created.

**Comment by hsd-dev** [comment]
Added the following lines in `fillRootElement`
```
if (notationModel.getSemanticElement() != null) {

   if (notationModel.getSemanticElement().getResolvedSemanticElement() == null) {
      SemanticElementReference reference = NotationFactory.eINSTANCE.createSemanticElementReference();
      reference.setElementId(model.getName());
      notationModel.getSemanticElement().setResolvedSemanticElement(reference);
   }
```
And now it works. Is this the correct way or is this more like a hack?

**Reply by hsd-dev** [reply]
this definitely seems like a hack, because every time the app starts or a new link is created, the nodes get repositioned to (0, 0) even thought the notation model contains the position information.

**Comment by ndoschek** [comment, marked as answer]
Hi @ipa-hsd!

Basically, the logic link between the semantic and the notation element is the `SemanticElementReference`.
To retrieve the notation information this reference is used to retrieve the position/size information.

In our `java-emf-theia` example, an id is used as element reference, in the following example the id **61bcc27a-3e93-4847-89e4-1d2433371b4b**:

Semantic model `example.tasklist`:
```
...
  <tasks id="61bcc27a-3e93-4847-89e4-1d2433371b4b" name="NewTask0"/>
...
```
Notation model `example.notation`:
```
...
  <elements xsi:type="notation:Shape">
    <semanticElement elementId="61bcc27a-3e93-4847-89e4-1d2433371b4b"/>
    <position x="382.0" y="101.0"/>
    <size width="67.0" height="56.0"/>
  </elements>
...
```

To be able to apply the notation data when the GModel is created, it is necessary that the semantic element reference can be properly resolved (see also the [`applyShapeData`](https://github.com/eclipse-glsp/glsp-server/blob/ffbd1b5588e91f711e64a968837e6a8b98ef21e1/plugins/org.eclipse.glsp.server.emf/src/org/eclipse/glsp/server/emf/notation/EMFNotationGModelFactory.java#L63) method).

So in your case, I assume that this reference cannot be properly resolved only by the name of the link.
This also explains why the shapes are positioned at `(0,0)` as the corresponding shape element cannot be resolved.

So I suggest you use the same approach by assigning an `id` via the `idGenerator` as it is done in the `java-emf-theia`'s [`CreateNodeTaskHandler`](https://github.com/eclipse-glsp/glsp-examples/blob/master/project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemf/handler/CreateTaskNodeHandler.java) (L#92).

HTH and best wishes,
Nina

**Reply by hsd-dev** [reply]
OK got it. I have now added `id` attribute to the `Links` element which is the root of my model. So the semantic model looks like
```
<?xml version="1.0" encoding="UTF-8"?>
<model:Links xmi:version="2.0" xmlns:xmi="http://www.omg.org/XMI" xmlns:model="http://www.ipa.fraunhofer.de/model" id="root" name="test">
  <body>
  </body>
</model:Links>
```
and `.notation`
```
<?xml version="1.0" encoding="ASCII"?>
<notation:Diagram xmi:version="2.0" xmlns:xmi="http://www.omg.org/XMI" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:notation="http://www.eclipse.org/glsp/notation">
  <semanticElement elementId="root"/>
</notation:Diagram>
```
But the resolved semantic element still prints `null` in `fillRootElement`
```java
System.out.println(notationModel.getSemanticElement());
// org.eclipse.glsp.server.emf.model.notation.impl.SemanticElementReferenceImpl@45633f5 (elementId: root)
System.out.println(notationModel.getSemanticElement().getResolvedSemanticElement());
// null
```
Now when I create a new link, the `id` disappears from the semantic model's root
```
<?xml version="1.0" encoding="UTF-8"?>
<model:Links xmi:version="2.0" xmlns:xmi="http://www.omg.org/XMI" xmlns:model="http://www.ipa.fraunhofer.de/model" name="test">
  <body>
    <link resolved="NewLink0" id="08cf29a0-ff7d-4921-bba1-4ac1df55df64"/>
  </body>
</model:Links>
```
and understably the `elementId` for each element in the notation model is empty
```
<?xml version="1.0" encoding="ASCII"?>
<notation:Diagram xmi:version="2.0" xmlns:xmi="http://www.omg.org/XMI" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:notation="http://www.eclipse.org/glsp/notation">
  <semanticElement elementId="root"/>
  <elements xsi:type="notation:Shape">
    <semanticElement/>
    <position x="533.0" y="173.0"/>
    <size width="60.0" height="25.0"/>
  </elements>
</notation:Diagram>
```
In the `java-emf-theia` I don't think the reference to the semantic element is set explicitly for the root, right?

In `CreateLinkNodeHandler`, the link id is printed as `null`
```java
Link newLink = createLink();
Command taskCommand = AddCommand.create(editingDomain, model.getBody(),
    ModelPackage.Literals.BODY__LINK, newLink);

System.out.println("link id: " + idGenerator.getOrCreateId(newLink));
// link id: null

protected Link createLink() {
    Link newLink = ModelFactory.eINSTANCE.createLink();
    setInitialName(newLink);
    newLink.setId(UUID.randomUUID().toString());
    return newLink;
}
```

**Reply by hsd-dev** [reply]
This file is not being generated for my Ecore model
```
<extension point="org.eclipse.emf.ecore.extension_parser">
  <!-- @generated model -->
  <parser
        type="model"
        class="org.eclipse.glsp.example.tasklist.model.util.ModelResourceFactoryImpl"/>
</extension>
```
Could that be the reason?

**Reply by hsd-dev** [reply]
Somehow the `idGenerator` is not working
```java
Link newLink = createLink();
System.out.println("link id: " + idGenerator.getOrCreateId(newLink) + ", " + newLink.getId());
// link id: null, 6b3be531-8782-481b-81fd-c6bac3215376
```
Maybe that's why the reference between notation and semantic element is not found?

**Reply by ndoschek** [reply]
Hi @ipa-hsd,

> But the resolved semantic element still prints null in fillRootElement
> 
> `System.out.println(notationModel.getSemanticElement());`
> `// org.eclipse.glsp.server.emf.model.notation.impl.SemanticElementReferenceImpl@45633f5 (elementId: root)`
> `System.out.println(notationModel.getSemanticElement().getResolvedSemanticElement());`
> `// null`

Regarding the resolved element: You could check the indexing process, set a breakpoint and see why the resolved element is not set/cannot be found here: [`createReference`](https://github.com/eclipse-glsp/glsp-server/blob/132e66d39b58b5de5ee359f9892fc35fce9e5800/plugins/org.eclipse.glsp.server.emf/src/org/eclipse/glsp/server/emf/notation/EMFNotationModelIndex.java#L150).

> This file is not being generated for my Ecore model

Yes, this file is necessary and should usually be generated within the code generation via the `.genmodel`. Did you change any properties in the GenModel perhaps?

> Somehow the idGenerator is not working

Did you bind the id generator correctly? E.g. as in the [`TaskListDiagramModule`](https://github.com/eclipse-glsp/glsp-examples/blob/54900faf011a935666330233799592f8c7ca4ea5/project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemf/TaskListDiagramModule.java#L54)?

> Maybe that's why the reference between notation and semantic element is not found?

Yes I assume so, if the ids keep disappearing, the resolving of semantic elements is probably not possible.

HTH, best wishes,

Nina

**Reply by hsd-dev** [reply]
`createReference` is never called for even the `java-emf-theia` example. The magic happens in [resolveDiagram](https://github.com/eclipse-glsp/glsp-server/blob/master/plugins/org.eclipse.glsp.server.emf/src/org/eclipse/glsp/server/emf/notation/EMFNotationModelIndex.java#L107). 

[This](https://github.com/eclipse-glsp/glsp-server/blob/master/plugins/org.eclipse.glsp.server.emf/src/org/eclipse/glsp/server/emf/notation/EMFSemanticIdConverter.java#L49) line returns `null` for my use-case, but `java-emf-theia` example correctly returns `root`. 

> Did you bind the id generator correctly? E.g. as in the [TaskListDiagramModule](https://github.com/eclipse-glsp/glsp-examples/blob/54900faf011a935666330233799592f8c7ca4ea5/project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemf/TaskListDiagramModule.java#L54)?

Yes 
```
@Override
protected Class<? extends EMFIdGenerator> bindEMFIdGenerator() {
  // all our elements inherit from Identifiable and have an ID attribute set
  return AttributeIdGenerator.class;
}

```

**Reply by hsd-dev** [reply]
```java
public class AttributeIdGenerator implements EMFIdGenerator {
   @Override
   public String getOrCreateId(final EObject element) {
      return EcoreUtil.getID(element);
   }
}
```
and in `EcoreUtil.class` 
```java
  public static java.lang.String getID(org.eclipse.emf.ecore.EObject eObject) {
    return null;
  }
```
I don't understand where is the correct implementation coming from for `java-emf-theia`.

**Reply by martin-fleck-at** [reply]
@ipa-hsd Which version of EMF are you using? For me, org.eclipse.emf.ecore.util.EcoreUtil.class (from org.eclipse.emf.ecore_2.26.0.v20220123-0838 - but I assume it stayed mostly the same through most versions) looks like this:
```
 public static String getID(EObject eObject)
 {
   EClass eClass = eObject.eClass();
   EAttribute eIDAttribute = eClass.getEIDAttribute();
   return eIDAttribute == null || !eObject.eIsSet(eIDAttribute) ? null : convertToString(
     eIDAttribute.getEAttributeType(),
     eObject.eGet(eIDAttribute));
 }
```

So for the attribute ID generator to work you would need to set an explicit ID attribute in your metamodel (e.g., setting the `iD="true"` in https://github.com/eclipse-glsp/glsp-examples/blob/master/project-templates/java-emf-theia/glsp-server/model/model.ecore#L7).

**Reply by hsd-dev** [reply]
Ah finally!! had missed setting it to true :D thanks a lot
